Saxon-JS performance comparisons

danielnaab opened the pull request:
Performance comparisons of Saxon-HE vs Saxon-JS usage. Implements #87

On my developer laptop, Saxon-HE performs better - roughly 3.7 vs 5.7 seconds.

This branch includes:

- A command-line interface for the UI code base to run a transform
- Updates to the Saxon-JS usage to support node.js - principally, using XPath instead of CSS selectors and delegating XML parsing to Saxon's platform implementations available at `SaxonJS.getPlatform()`
- Updates to `validate_with_schematron.sh` to support skipping Schematron compilation, for the sake of isolating the transform step for timing purposes

ohsh6o commented on resources/validations/ui/src/context/cli/index.ts:
```diff
@@ -0,0 +1,23 @@
+});
+
+controller.parse(process.argv);
+console.log(controller.opts());
```
NB: Question, when running the application as instructed is this what emits `{}` before the validations count?

```sh
$ time npm run cli  "validate" ../test/demo/FedRAMP-SSP-OSCAL-Template.xml 

> cli
> ts-node src/context/cli/index.ts "validate" "../test/demo/FedRAMP-SSP-OSCAL-Template.xml"

{}
Found 46 assertions

real    0m5.744s
user    0m7.012s
sys     0m0.365s
```

ohsh6o commented on resources/validations/ui/src/context/cli/index.ts:
```diff
@@ -0,0 +1,23 @@
+const controller = CommandLineController({
+  readStringFile,
+  validateSchematron: SaxonJsSchematronValidatorGateway({
+    sefUrl: `file:///Users/dan/src/10x/fedramp-automation/resources/validations/ui/public/ssp.sef.json`,
```
B: I am not sure how to do this with Typescript when it is compiled down, but can we set up the equivalent of the `__dirname` functionality in Javascript.

ohsh6o commented on resources/validations/ui/src/context/cli/index.ts:
```diff
@@ -0,0 +1,23 @@
+    // The npm version of saxon-js is for node; currently, we load the
+    // browser version via a script tag in index.html.
+    SaxonJS: SaxonJS,
+    baselinesBaseUrl: `file:///Users/dan/src/10x/fedramp-automation/baselines/rev4/xml`,
+    registryBaseUrl: `file:///Users/dan/src/10x/fedramp-automation/resources/xml`,
```
B: Ditto here, update paths.

ohsh6o commented on resources/validations/ui/src/context/shared/saxon-js-gateway.test.ts:
```diff
@@ -0,0 +1,40 @@
+
+import { SaxonJsSchematronValidatorGateway } from './saxon-js-gateway';
+
+describe('saxon-js gateway', () => {
```
NB: Thanks for adding tests.

ohsh6o reviewed (changes requested):
Some initial change recommendations.

- Explicit in the current AC we are to measure "the Java version of the validation pipeline." Will this be covered in a subsequent PR or something we can achieve in this sprint?
- Also, for consideration, and maybe it is not clear in the AC, are we going to measure this in CI to know if any changes negatively impact performance?

danielnaab commented:
> * Explicit in the current AC we are to measure "the Java version of the validation pipeline." Will this be covered in a subsequent PR or something we can achieve in this sprint?

I had updated the shell script to isolate the transform step, and added a note to the README on how to run it here: https://github.com/18F/fedramp-automation/blob/performance/resources/validations/ui/README.md#saxon-he

> * Also, for consideration, and maybe it is not clear in the AC, are we going to measure this in CI to know if any changes negatively impact performance?

I'd have to think about how to handle that - were you thinking CI would fail if a threshold was exceeded? Or another way to surface the speed?

ohsh6o commented:
> I'd have to think about how to handle that - were you thinking CI would fail if a threshold was exceeded? Or another way to surface the speed?

I was thinking we record the speed for now, measure the difference up or down, and review during PR reviews and/or sprint reviews for significant deltas, do not block CI builds when slow at this time.

Sorry for the delayed response, @danielnaab.

ohsh6o commented:
> > * Explicit in the current AC we are to measure "the Java version of the validation pipeline." Will this be covered in a subsequent PR or something we can achieve in this sprint?
> 
> I had updated the shell script to isolate the transform step, and added a note to the README on how to run it here: https://github.com/18F/fedramp-automation/blob/performance/resources/validations/ui/README.md#saxon-he

Sorry, per sprint review, I missed this part of the convo and we will accept this as part of the work, and the XSLT baselines profiling we will move out into another story.

ohsh6o reviewed (approved):
LGTM.
